redwood-layout-view/src/main/kotlin/app/cash/redwood/layout/view/YogaLayout.kt

```diff
@@ -10,88 +10,100 @@ import android.annotation.SuppressLint
 import android.content.Context
 import android.view.View
 import android.view.ViewGroup
+import app.cash.redwood.layout.api.Constraint
 import app.cash.redwood.yoga.Node
 import app.cash.redwood.yoga.Size
 import kotlin.math.roundToInt
 
 @SuppressLint("ViewConstructor")
 internal class YogaLayout(context: Context) : ViewGroup(context) {
   val rootNode = Node()
+    .apply {
+      this.context = this@YogaLayout
+    }
+
+  internal var widthConstraint = Constraint.Wrap
+  internal var heightConstraint = Constraint.Wrap
 
   private fun applyLayout(node: Node, xOffset: Float, yOffset: Float) {
-    val view = node.view
-    if (view != null && view !== this) {
+    val view = node.context as View
+    if (view !== this) {
       if (view.visibility == GONE) return
 
-      val width = node.width.roundToInt()
-      val height = node.height.roundToInt()
-      val widthSpec = MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY)
-      val heightSpec = MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY)
-      view.measure(widthSpec, heightSpec)
-
       val left = (xOffset + node.left).roundToInt()
       val top = (yOffset + node.top).roundToInt()
-      val right = left + view.measuredWidth
-      val bottom = top + view.measuredHeight
+      val right = left + node.width.roundToInt()
+      val bottom = top + node.height.roundToInt()
+
+      // We already know how big we want this view to be. But we measure it to trigger side-effects
+      // that the view needs to render itself correctly. In particular, `TextView` needs this
+      // otherwise it won't apply gravity correctly.
+      val width = right - left
+      val height = bottom - top
+      view.measure(
+        MeasureSpec.makeMeasureSpec(width, MeasureSpec.EXACTLY),
+        MeasureSpec.makeMeasureSpec(height, MeasureSpec.EXACTLY),
+      )
```

Comment:
Today I learned that `measure()` keeps a cache so it’s almost a no-op to call it if you’ve just called it with the same parameters.